### ISSUE TITLE
What are the validity requirements of wide pointers/references with `dyn Trait` tail?

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#516]
This is the last remaining part of https://github.com/rust-lang/unsafe-code-guidelines/issues/166, now that https://github.com/rust-lang/unsafe-code-guidelines/issues/510 has been resolved.
Note that the *safety* invariant for both of these is pretty much set already since casts from `*const dyn Trait` to `*const dyn Supertrait` are safe -- that means the vtable pointer must be valid for the given trait, even on raw pointers.

For the validity of the metadata in references, there seems to be general consensus that it is the same as the safety invariant: the vtable must be a vtable for the right trait (and an arbitrary type).

So the open question is what to do with the validity of metadata in raw pointers. The two contradicting goals pursued by different people here are:
- avoid the confusion that can arise when validity and safety invariant diverge
- avoid the pitfall of a NULL vtable pointer (or some other sentinel value) being insta-UB

This was [discussed in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Validity.20of.20wide.20pointer.2Freference.20metadata) somewhat recently. My proposal was to take a maximally liberal validity invariant, and allow any vtable pointer, but several people had concerns that it would be confusing to allow allow a null (or general, unsafe) vtable to exist but then make it UB to cast that pointer or use it as a dyn receiver, even though that operation is safe. One possible compromise was to explicitly allow only null and then make that actually safe by adding checks on casts and dyn calls -- but that would go against the idea of using raw pointers in cases where performance matters and safe code has too many checks.

It is also worth noting that t-lang has an FCP that touches on this question in https://github.com/rust-lang/rust/issues/101336, where they recommend that the validity invariant should require vtable pointers to be *word-aligned and non-null but nothing else*. This would be half-way between "maximally liberal" and the expected invariant for `&dyn Trait`; it raises the question whether `&dyn Trait` should be made to have the same validity invariant just to avoid potential confusion from the difference.

**Comment by RalfJung** [comment]
One interesting sub-question to consider here is the exact definition of "the right trait". Currently in Miri, "the right trait" is defined as "has the same *principal* trait", where the "principal" trait is the trait that is left after removing all auto traits.

So for instance, transmutes between `dyn Debug` and `dyn Debug+Send` are allowed, but transmutes between `dyn Debug` and `dyn Display` are not. `dyn Debug + Send` vs `dyn Send` is also not allowed, because the former has `Debug` as the principal trait while the latter does not have a principal trait.

It seems to me that the most consistent and conservative option right now would be to require full equality of the trait list, i.e. to also rule out  transmutes between `dyn Debug` and `dyn Debug+Send`.

**Comment by RalfJung** [comment]
vtables are generated by the compiler. It is UB to point a vtable pointer to anything but a compiler-generated vtable. So it is impossible to even have a vtable that violates that property.

**Comment by celinval** [comment]
Ah, I just noticed this statement in the Rust reference:

> dyn Trait metadata must be a pointer to a compiler-generated vtable for Trait.

Thanks!

**Comment by celinval** [comment]
Actually, the reference states that the requirement for raw pointers is subject to debate:

> - dyn Trait metadata must be a pointer to a compiler-generated vtable for Trait. (For raw pointers, this requirement remains a subject of some debate.)

But is it?

**Comment by RalfJung** [comment]
Yes, and this thread is part of the debate. :)

**Comment by VorpalBlade** [comment]
What are the concrete use cases of having non-compiler-generated vtables? The main case I can think of is if another language (or implementation of Rust) wants to interoperate with Rust.

But without a stable ABI that doesn't work for a lot of other reasons anyway. So given that FFI isn't a realistic reason (at least not for quite a long time), what are the other possible use cases? I suspect the existance or lack of such use cases could help inform what Rust wants to do when it comes to vtables.

**Comment by RalfJung** [comment]
I have updated the OP based on @nikomatsakis's [comment here](https://github.com/rust-lang/reference/pull/1622#issuecomment-2580815386). I didn't quite remember that t-lang actually had an FCP that mentions a concrete validity invariant, requiring vtable pointers to be *word-aligned and non-null but nothing else.*

I would find it strange to have such a non-trivial validity invariant for raw pointers but then have something different again for references. So if this is what we go for, IMO we should then use this for both vtable metadata in both references and raw pointers.

**Comment by nikomatsakis** [comment]
Consistency would be expected yes

**Comment by Darksonn** [comment]
It has come to my attention that `derive(CoercePointee)` interacts with this. Specifically, the feature lets you trigger UB if you can obtain a raw pointer whose vtable is not a valid vtable for the type of the raw pointer. This is because the feature can be used to safely dispatch on the raw pointer's vtable.

See this for an example of unsoundness: https://github.com/rust-lang/rust/issues/127323#issuecomment-2642989675

**Comment by RalfJung** [comment]
That's not new, you can already trigger UB via dyn upcasts that way. But that only means that the *library*/"safety" invariant of raw pointers needs to require vtable validity; it does not decide the *language*/"validity" invariant.

**Comment by Darksonn** [comment]
Yeah, sorry I missed that this issue is about validity invariants and not safety invariants.

**Comment by RalfJung** [comment]
I entirely forgot that this is still open... so we aren't done with validity invariants after all @rust-lang/opsem 🙈 .

What are the general vibes for this question? I used to prefer the very liberal invariant, but at this point I'd be willing to go with almost anything in order to get the question settled. ;) In other words, I don't care very strongly. Here are the options I am aware of, loosely from most strict to least strict:
1. The strictest possible requirement is "must be a valid vtable for the trait". This is almost exactly what references require (references additionally require the total size to fit into `isize`, but that's more a consequence of "references are dereferenceable" than a separate requirement). This is what we currently document and what Miri implements since [two years ago](https://github.com/rust-lang/rust/pull/124220) (and I don't recall a single issue being filed or question on Zulip about this, adding this check seems to have gone quite smoothly). As the maximal-UB option it's also the most future-compatible, we can always remove UB later.
2. There's a [t-lang FCP](https://github.com/rust-lang/rust/issues/101336) that [*recommends*](https://github.com/rust-lang/rust/issues/101336#issuecomment-1242719874) "must be word-aligned and non-null" (and nothing else). I honestly see no good reason to pick this option; at that point saying that the vtable must also match the trait loses us basically nothing while gaining us consistency with the requirement for references. I was not able to really figure out the rationale for this choice. It seems an alternative they considered was "vtable must match the hidden actual type of the pointed-to value", which is a meaningless definition as there may be no pointed-to value, so it was discarded. The option of "vtable must match pointee type" was [mentioned by me](https://github.com/rust-lang/rust/issues/101336#issuecomment-1240944016) but apparently not discussed.
3. There's some desire to allow zero-initializing all raw pointers, so we could say "must be null or a valid vtable for the trait". This loses us a null niche (but preserves the "word-aligned" niche as we know that all valid vtables are word-aligned, and anyway I don't think niches on raw pointers are so important). It allows zero-initializing, *but the resulting value is not safe*. It is similar to a `&T` where the memory is not valid for `T` -- it's a value you can use locally inside your library, but you must not hand this value to outside code. I used to think zero-initializing would be important, but it doesn't seem *that* important given the lack of fallout from the Miri change. I guess the kind of code that has to zero-initialize things doesn't often deal with `*const dyn Trait` pointers. (Also, there is an alternative way to achieve zero-initializing: we could guarantee that `Option<*const dyn Trait>` allows zero-initializing and uses that to represent `None`.)
4. And finally we could just say that the vtable point is itself basically just a raw pointer, and can point to anything. This obviously loses all niches, and has the same problem as the previous option regarding "valid but not safe" values.

My personal opinion is that I slightly prefer 4 over 1, but would tick my box for either choice. I consider 2 and 3 to be awkward middle grounds that are not worth the confusion they may cause.

@CAD97 you seemed to argue quite strongly in that old t-lang FCP for a liberal validity requirement. What are your thoughts on this today?

@saethlin would be interesting whether your crater-at-home runs uncovered notable amounts of invalid vtables in raw pointers.

**Comment by saethlin** [comment]
> [@saethlin](https://github.com/saethlin) would be interesting whether your crater-at-home runs uncovered notable amounts of invalid vtables in raw pointers. 

I did a quick check, and I have not yet found any crates which are constructing invalid `*const dyn Trait` but I did find a handful that are constructing invalid `Box<dyn Trait>` by transmuting for example `Box<dyn FnOnce()>` to `Box<dyn FnOnce() -> &str>`.

Seems possible there might be a selection effect where people who learn those transmutes aren't valid by using Miri are then steered away from just replacing it with a raw dyn Trait pointer.
